providers/config.go

```diff
@@ -0,0 +1,100 @@
+package providers
+
+import (
+	"errors"
+	"fmt"
+	"html/template"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	// configFileLoc is the name of the config file.
+	configFileRelativeLoc = "providers.yaml"
+)
+
+var (
+	ErrProviderConfigNotFound = errors.New("provider config not found")
+)
+
+// Config is the entire configuration for all providers.
+type Config struct {
+	Providers map[Provider]map[string]string `yaml:"providers"`
+}
+
+// ReadConfig reads the configuration from the config file for
+// a specific provider. It also performs string substitution
+// on the values in the config that are surrounded by {{}}.
+// The provider YAML has more details on how it works.
+func ReadConfig(provider Provider, substitutions map[string]string) (map[string]string, error) {
```

Comment:
It would be good if this actually returned a Go type rather than a map of strings, then the calling code can work with the returned result more easily.

Reply:
I played around with this, but configuration for providers will diverge, leaving us with two options - 
* Define config structs for each provider, which is a lot of work 
* Use a map (optionally define keys as constants with which to look the map up)

Or we could do something in between where we define some static configuration that is a struct, and some per-provider config that is a map.

I went with a map because the person adding a new config key will be the one using it so it should be okay as long as we are careful & we keep extending the testcases. Secondly, this is something that will not change often.